File: ruby/test/onramp_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class OnrampTest < Minitest::Test
  GATEWAY = "https://app.datagrout.ai"
  ONRAMP_URL = "#{GATEWAY}/onramp"
  COMPLETE_URL = "#{GATEWAY}/onramp/complete"
  TOKEN_URL = "#{GATEWAY}/servers/abc/oauth/token"

  INIT_RESPONSE = { "session_token" => "sess_abc123" }.freeze

  COMPLETE_RESPONSE = {
    "client_id" => "agt_abc123",
    "client_secret" => "sk_xyz789",
    "token_url" => TOKEN_URL,
    "mcp_url" => "#{GATEWAY}/servers/abc/mcp",
    "rpc_url" => "#{GATEWAY}/servers/abc/rpc",
    "scopes" => ["mcp:read", "tools:call"],
    "expires_in" => 2_592_000
  }.freeze

  TOKEN_RESPONSE = { "access_token" => "tok_live123" }.freeze

  def setup
    WebMock.disable_net_connect!
  end

  def teardown
    WebMock.reset!
    WebMock.allow_net_connect!
  end

  def default_opts
    DatagroutConduit::Onramp::OnrampOptions.new(
      gateway: GATEWAY,
      agent_name: "test-agent",
      agent_type: "claude-sonnet-4-6",
      intended_use: "Testing."
    )
  end

  # ─── OnrampOptions ──────────────────────────────────────────────────────────

  def test_onramp_options_required_fields
    opts = DatagroutConduit::Onramp::OnrampOptions.new(
      gateway: GATEWAY,
      agent_name: "my-agent"
    )
    assert_equal GATEWAY, opts.gateway
    assert_equal "my-agent", opts.agent_name
    assert_nil opts.agent_type
    assert_nil opts.intended_use
    assert_nil opts.access_code
  end

  def test_onramp_options_all_fields
    opts = DatagroutConduit::Onramp::OnrampOptions.new(
      gateway: GATEWAY,
      agent_name: "my-agent",
      agent_type: "gpt-4o",
      intended_use: "extraction",
      access_code: "code123"
    )
    assert_equal "gpt-4o", opts.agent_type
    assert_equal "extraction", opts.intended_use
    assert_equal "code123", opts.access_code
  end

  # ─── OnrampCredentials ──────────────────────────────────────────────────────

  def test_onramp_credentials_fields
    creds = DatagroutConduit::Onramp::OnrampCredentials.new(
      client_id: "agt_abc",
      client_secret: "sk_xyz",
      token_url: TOKEN_URL,
      scopes: ["mcp:read"],
      expires_in: 2_592_000,
      mcp_url: "#{GATEWAY}/servers/abc/mcp",
      rpc_url: "#{GATEWAY}/servers/abc/rpc"
    )
    assert_equal "agt_abc", creds.client_id
    assert_equal ["mcp:read"], creds.scopes
    assert_equal 2_592_000, creds.expires_in
  end

  def test_onramp_credentials_nil_urls
    creds = DatagroutConduit::Onramp::OnrampCredentials.new(
      client_id: "x", client_secret: "y", token_url: TOKEN_URL,
      scopes: [], expires_in: 0
    )
    assert_nil creds.mcp_url
    assert_nil creds.rpc_url
  end

  # ─── register_only ──────────────────────────────────────────────────────────

  def test_register_only_sends_init_then_complete
    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )

    stub_request(:post, COMPLETE_URL)
      .with(headers: { "Authorization" => "Bearer sess_abc123" })
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(COMPLETE_RESPONSE)
      )

    creds = DatagroutConduit::Onramp.register_only(default_opts)

    assert_equal "agt_abc123", creds.client_id
    assert_equal "sk_xyz789", creds.client_secret
    assert_equal TOKEN_URL, creds.token_url
    assert_equal ["mcp:read", "tools:call"], creds.scopes
    assert_equal "#{GATEWAY}/servers/abc/mcp", creds.mcp_url
    assert_equal 2_592_000, creds.expires_in
  end

  def test_register_only_includes_optional_fields
    stub_request(:post, ONRAMP_URL)
      .with(body: hash_including(
        "agent_name" => "test-agent",
        "agent_type" => "claude-sonnet-4-6",
        "intended_use" => "Testing."
      ))
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )

    stub_request(:post, COMPLETE_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(COMPLETE_RESPONSE)
      )

    DatagroutConduit::Onramp.register_only(default_opts)
    assert_requested :post, ONRAMP_URL
  end

  def test_register_only_omits_nil_optional_fields
    bare_opts = DatagroutConduit::Onramp::OnrampOptions.new(
      gateway: GATEWAY,
      agent_name: "bare"
    )

    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )

    stub_request(:post, COMPLETE_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(COMPLETE_RESPONSE)
      )

    DatagroutConduit::Onramp.register_only(bare_opts)

    assert_not_requested :post, ONRAMP_URL,
      body: hash_including("agent_type" => /\S+/)
  end

  def test_register_only_raises_on_init_rejected
    stub_request(:post, ONRAMP_URL).to_return(status: 429, body: "rate_limited")

    assert_raises DatagroutConduit::Onramp::OnrampError do
      DatagroutConduit::Onramp.register_only(default_opts)
    end
  end

  def test_register_only_raises_on_complete_rejected
    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )
    stub_request(:post, COMPLETE_URL).to_return(status: 410, body: "expired")

    assert_raises DatagroutConduit::Onramp::OnrampError do
      DatagroutConduit::Onramp.register_only(default_opts)
    end
  end

  def test_register_only_handles_absent_mcp_url
    partial = COMPLETE_RESPONSE.reject { |k, _| k == "mcp_url" || k == "rpc_url" }

    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )
    stub_request(:post, COMPLETE_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(partial)
      )

    creds = DatagroutConduit::Onramp.register_only(default_opts)
    assert_nil creds.mcp_url
    assert_nil creds.rpc_url
  end

  # ─── register_and_exchange ──────────────────────────────────────────────────

  def test_register_and_exchange_returns_creds_and_token
    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )
    stub_request(:post, COMPLETE_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(COMPLETE_RESPONSE)
      )
    stub_request(:post, TOKEN_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(TOKEN_RESPONSE)
      )

    creds, token = DatagroutConduit::Onramp.register_and_exchange(default_opts)

    assert_equal "agt_abc123", creds.client_id
    assert_equal "tok_live123", token
  end

  def test_register_and_exchange_raises_on_token_exchange_failure
    stub_request(:post, ONRAMP_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(INIT_RESPONSE)
      )
    stub_request(:post, COMPLETE_URL)
      .to_return(
        status: 200,
        headers: { "Content-Type" => "application/json" },
        body: JSON.generate(COMPLETE_RESPONSE)
      )
    stub_request(:post, TOKEN_URL).to_return(status: 401, body: "invalid_client")

    assert_raises DatagroutConduit::Onramp::OnrampError do
      DatagroutConduit::Onramp.register_and_exchange(default_opts)
    end
  end
end

File: elixir/test/onramp_test.exs
defmodule DatagroutConduit.OnrampTest do
  use ExUnit.Case, async: true

  alias DatagroutConduit.Onramp
  alias DatagroutConduit.Onramp.{OnrampOptions, OnrampCredentials}

  @gateway "https://app.datagrout.ai"
  @token_url "#{@gateway}/servers/abc/oauth/token"

  @complete_body %{
    "client_id" => "agt_abc123",
    "client_secret" => "sk_xyz789",
    "token_url" => @token_url,
    "mcp_url" => "#{@gateway}/servers/abc/mcp",
    "rpc_url" => "#{@gateway}/servers/abc/rpc",
    "scopes" => ["mcp:read", "tools:call"],
    "expires_in" => 2_592_000
  }

  def default_opts do
    %OnrampOptions{
      gateway: @gateway,
      agent_name: "test-agent",
      agent_type: "claude-sonnet-4-6",
      intended_use: "Testing."
    }
  end

  # ─── OnrampOptions struct ─────────────────────────────────────────────────────

  describe "OnrampOptions" do
    test "holds required fields" do
      opts = %OnrampOptions{gateway: @gateway, agent_name: "my-agent"}
      assert opts.gateway == @gateway
      assert opts.agent_name == "my-agent"
    end

    test "optional fields default to nil" do
      opts = %OnrampOptions{gateway: @gateway, agent_name: "bare"}
      assert is_nil(opts.agent_type)
      assert is_nil(opts.intended_use)
      assert is_nil(opts.access_code)
    end

    test "all fields can be set" do
      opts = %OnrampOptions{
        gateway: @gateway,
        agent_name: "my-agent",
        agent_type: "gpt-4o",
        intended_use: "extraction",
        access_code: "code123"
      }
      assert opts.agent_type == "gpt-4o"
      assert opts.access_code == "code123"
    end
  end

  # ─── OnrampCredentials struct ─────────────────────────────────────────────────

  describe "OnrampCredentials" do
    test "holds all fields" do
      creds = %OnrampCredentials{
        client_id: "agt_abc",
        client_secret: "sk_xyz",
        token_url: @token_url,
        scopes: ["mcp:read"],
        expires_in: 2_592_000,
        mcp_url: "#{@gateway}/servers/abc/mcp",
        rpc_url: "#{@gateway}/servers/abc/rpc"
      }
      assert creds.client_id == "agt_abc"
      assert creds.scopes == ["mcp:read"]
      assert creds.expires_in == 2_592_000
    end

    test "mcp_url and rpc_url default to nil" do
      creds = %OnrampCredentials{
        client_id: "x", client_secret: "y", token_url: @token_url
      }
      assert is_nil(creds.mcp_url)
      assert is_nil(creds.rpc_url)
    end

    test "scopes defaults to empty list" do
      creds = %OnrampCredentials{client_id: "x", client_secret: "y", token_url: @token_url}
      assert creds.scopes == []
    end
  end

  # ─── register_only ────────────────────────────────────────────────────────────

  describe "register_only/1" do
    test "returns {:ok, creds} on successful two-step flow" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        cond do
          conn.request_path == "/onramp" ->
            Req.Test.json(conn, %{"session_token" => "sess_abc123"})

          conn.request_path == "/onramp/complete" ->
            Req.Test.json(conn, @complete_body)
        end
      end)

      assert {:ok, creds} = Onramp.register_only(default_opts())
      assert creds.client_id == "agt_abc123"
      assert creds.client_secret == "sk_xyz789"
      assert creds.mcp_url == "#{@gateway}/servers/abc/mcp"
      assert creds.scopes == ["mcp:read", "tools:call"]
      assert creds.expires_in == 2_592_000
    end

    test "returns {:error, ...} when init is rejected" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        Plug.Conn.send_resp(conn, 429, "rate_limited")
      end)

      assert {:error, {:onramp_init_rejected, 429, _}} = Onramp.register_only(default_opts())
    end

    test "returns {:error, ...} when complete is rejected" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        cond do
          conn.request_path == "/onramp" ->
            Req.Test.json(conn, %{"session_token" => "sess_abc123"})

          conn.request_path == "/onramp/complete" ->
            Plug.Conn.send_resp(conn, 410, "expired")
        end
      end)

      assert {:error, {:onramp_complete_rejected, 410, _}} = Onramp.register_only(default_opts())
    end

    test "handles absent mcp_url and rpc_url gracefully" do
      partial = Map.drop(@complete_body, ["mcp_url", "rpc_url"])

      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        cond do
          conn.request_path == "/onramp" ->
            Req.Test.json(conn, %{"session_token" => "sess_abc123"})

          conn.request_path == "/onramp/complete" ->
            Req.Test.json(conn, partial)
        end
      end)

      assert {:ok, creds} = Onramp.register_only(default_opts())
      assert is_nil(creds.mcp_url)
      assert is_nil(creds.rpc_url)
    end
  end

  # ─── register_and_exchange/1 ──────────────────────────────────────────────────

  describe "register_and_exchange/1" do
    test "returns {:ok, {creds, token}} on full success" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        cond do
          conn.request_path == "/onramp" ->
            Req.Test.json(conn, %{"session_token" => "sess_abc123"})

          conn.request_path == "/onramp/complete" ->
            Req.Test.json(conn, @complete_body)

          String.ends_with?(conn.request_path, "/oauth/token") ->
            Req.Test.json(conn, %{"access_token" => "tok_live123"})
        end
      end)

      assert {:ok, {creds, token}} = Onramp.register_and_exchange(default_opts())
      assert creds.client_id == "agt_abc123"
      assert token == "tok_live123"
    end

    test "propagates init error" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        Plug.Conn.send_resp(conn, 429, "rate_limited")
      end)

      assert {:error, {:onramp_init_rejected, 429, _}} =
               Onramp.register_and_exchange(default_opts())
    end

    test "propagates token exchange error" do
      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        cond do
          conn.request_path == "/onramp" ->
            Req.Test.json(conn, %{"session_token" => "sess_abc123"})

          conn.request_path == "/onramp/complete" ->
            Req.Test.json(conn, @complete_body)

          true ->
            Plug.Conn.send_resp(conn, 401, "invalid_client")
        end
      end)

      assert {:error, {:token_exchange_failed, 401, _}} =
               Onramp.register_and_exchange(default_opts())
    end
  end

  # ─── exchange_token/1 ─────────────────────────────────────────────────────────

  describe "exchange_token/1" do
    test "returns {:ok, token} on success" do
      creds = %OnrampCredentials{
        client_id: "agt_abc",
        client_secret: "sk_xyz",
        token_url: @token_url
      }

      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        Req.Test.json(conn, %{"access_token" => "tok_live123"})
      end)

      assert {:ok, "tok_live123"} = Onramp.exchange_token(creds)
    end

    test "returns {:error, ...} on failure" do
      creds = %OnrampCredentials{
        client_id: "x", client_secret: "y", token_url: @token_url
      }

      Req.Test.stub(DatagroutConduit.Onramp, fn conn ->
        Plug.Conn.send_resp(conn, 401, "invalid_client")
      end)

      assert {:error, {:token_exchange_failed, 401, _}} = Onramp.exchange_token(creds)
    end
  end
end
